sdk/core/Azure.Core/src/Serialization/ModelSerializer.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Reflection;
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// Serializer class for Azure models.
+    /// </summary>
+    public static class ModelSerializer
+    {
+        /// <summary>
+        /// Serialize a model.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="model"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public static BinaryData Serialize<T>(T model, ModelSerializerOptions? options = default) where T : class, IModelSerializable
+        {
+            // if options.Serializers is set and the model is in the dictionary, use the serializer
+            if (options != null)
+            {
+                ObjectSerializer? serializer;
+
+                if (options.Serializers.TryGetValue(typeof(T), out serializer))
+                {
+                    return serializer.Serialize(model);
+                }
+            }
+            return model.Serialize(options ?? new ModelSerializerOptions());
+        }
+
+        /// <summary>
+        /// Serialize a XML model. Todo: collapse this method when working - need compile check over runtime
+        /// </summary>
+        /// <returns></returns>
+        public static T Deserialize<T>(BinaryData data, ModelSerializerOptions? options = default) where T : class, IModelSerializable
+        {
+            return (T)Deserialize(data, typeof(T), options);
```

Comment:
Is there a way to make this so it wouldn't box a value type?  Or is that irrelevant once we've taken on the overhead of serialization?

Reply:
Do you mean if someone calls `Deserialize<int>` it would box it?  I don't think this is possible given the where constraint on T.

Reply:
Ah, I missed the constraint - you are right.  Thanks!